packages/api/src/perma-cache/post.js

```diff
@@ -0,0 +1,238 @@
+/* eslint-env serviceworker, browser */
+/* global Response */
+
+import { FAR_FUTURE, PAD_LEN, MAX_ALLOWED_URL_LENGTH } from '../constants.js'
+import { InvalidUrlError, TimeoutError, HTTPError } from '../errors.js'
+import { JSONResponse } from '../utils/json-response.js'
+import { normalizeCid } from '../utils/cid.js'
+
+/**
+ * @typedef {import('../env').Env} Env
+ * @typedef {{ userId: string, r2Key: string, date: string }} Key
+ *
+ * @typedef {Object} IpfsUrlParts
+ * @property {string} cid
+ * @property {string} url
+ * @property {string} path
+ */
+
+/**
+ * Handle perma-cache post request
+ *
+ * @param {Request} request
+ * @param {Env} env
+ */
+export async function permaCachePost(request, env) {
+  const sourceUrl = getSourceUrl(request, env)
+  const normalizedUrl = getNormalizedUrl(sourceUrl, env)
+  const r2Key = normalizedUrl.toString()
+
+  // Validate if URL is not already perma cached by user
+  const kvPrefix = `${request.auth.user.id}:${r2Key}`
+  const { keys } = await env.PERMACACHE.list({
+    prefix: kvPrefix,
+  })
+  if (keys.length > 0) {
+    throw new HTTPError('The provided url was already perma cached', 400)
+  }
+
+  // Validate if we already have it in R2
+  let r2Object
+  try {
+    r2Object = await env.SUPERHOT.head(normalizedUrl.toString())
+  } catch (err) {
+    // TODO: R2 currently throws error here and does not follow types to return undefined.
+    // They say it will change soon to return undefined...
+    console.log(err)
+  }
+
+  if (!r2Object) {
+    // Fetch Response from provided URL
+    const response = await getResponse(request, env, normalizedUrl)
+    if (!response.ok) {
+      throw new HTTPError(await response.text(), 400)
+    }
+
+    // TODO: Validate headers per https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control
+    // Store in R2 and add to Database if not existent
+    r2Object = await env.SUPERHOT.put(r2Key, response.body, {
+      httpMetadata: response.headers,
```

Comment:
```suggestion
      httpHeaders: response.headers,
```

Reply:
We can't choose it 😅  https://github.com/cloudflare/workers-types/blob/master/index.d.ts#L1122